drivers/serial/usart_gd32.c

```diff
@@ -93,25 +148,201 @@ static int usart_gd32_err_check(const struct device *dev)
+void usart_gd32_irq_callback_set(const struct device *dev,
+				uart_irq_callback_user_data_t cb,
+				void *user_data)
+{
+	struct gd32_usart_data *const data = dev->data;
+
+	data->user_cb = cb;
+	data->user_data = user_data;
+}
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
 static const struct uart_driver_api usart_gd32_driver_api = {
 	.poll_in = usart_gd32_poll_in,
 	.poll_out = usart_gd32_poll_out,
 	.err_check = usart_gd32_err_check,
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.fifo_fill = usart_gd32_fifo_fill,
+	.fifo_read = usart_gd32_fifo_read,
+	.irq_tx_enable = usart_gd32_irq_tx_enable,
+	.irq_tx_disable = usart_gd32_irq_tx_disable,
+	.irq_tx_ready = usart_gd32_irq_tx_ready,
+	.irq_tx_complete = usart_gd32_irq_tx_complete,
+	.irq_rx_enable = usart_gd32_irq_rx_enable,
+	.irq_rx_disable = usart_gd32_irq_rx_disable,
+	.irq_rx_ready = usart_gd32_irq_rx_ready,
+	.irq_err_enable = usart_gd32_irq_err_enable,
+	.irq_err_disable = usart_gd32_irq_err_disable,
+	.irq_is_pending = usart_gd32_irq_is_pending,
+	.irq_update = usart_gd32_irq_update,
+	.irq_callback_set = usart_gd32_irq_callback_set,
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
 };
 
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+#define GD32_USART_IRQ_HANDLER_DECL(n)				\
+	static void usart_gd32_config_func_##n(const struct device *dev)
+#define GD32_USART_IRQ_HANDLER_FUNC(n)				\
+	.irq_config_func = usart_gd32_config_func_##n
+#define GD32_USART_IRQ_HANDLER(n)				\
+	static void usart_gd32_config_func_##n(const struct device *dev)			\
+	{								       \
+		IRQ_CONNECT(DT_INST_IRQN(n),				\
+			DT_INST_IRQ(n, priority),				\
+			usart_gd32_isr, DEVICE_DT_INST_GET(n),		\
+			0);							\
+		irq_enable(DT_INST_IRQN(n));				\
+	}
+#else /* CONFIG_UART_INTERRUPT_DRIVEN */
+#define GD32_USART_IRQ_HANDLER_DECL(n)
+#define GD32_USART_IRQ_HANDLER_FUNC(n)
+#define GD32_USART_IRQ_HANDLER(n)
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
 #define GD32_USART_INIT(n)							\
-	static struct gd32_usart_data usart##n##_gd32_data = {			\
+	GD32_USART_IRQ_HANDLER_DECL(n);					\
+	static struct gd32_usart_data usart_gd32_data_##n = {			\
 		.baud_rate = DT_INST_PROP(n, current_speed),			\
 	};									\
-	static const struct gd32_usart_config usart##n##_gd32_config = {	\
+	static const struct gd32_usart_config usart_gd32_config_##n = {	\
 		.reg = DT_INST_REG_ADDR(n),					\
 		.rcu_periph_clock = DT_INST_PROP(n, rcu_periph_clock),		\
+		.parity = DT_ENUM_IDX_OR(DT_DRV_INST(index), parity, UART_CFG_PARITY_NONE),	\
+		 GD32_USART_IRQ_HANDLER_FUNC(n)					\
 	};									\
 	DEVICE_DT_INST_DEFINE(n, &usart_gd32_init,				\
 			      NULL,						\
-			      &usart##n##_gd32_data,				\
-			      &usart##n##_gd32_config, PRE_KERNEL_1,		\
-			      CONFIG_KERNEL_INIT_PRIORITY_DEVICE,		\
-			      &usart_gd32_driver_api);
+			      &usart_gd32_data_##n,				\
+			      &usart_gd32_config_##n, PRE_KERNEL_1,		\
+			      CONFIG_SERIAL_INIT_PRIORITY,		\
+			      &usart_gd32_driver_api);			\
+	GD32_USART_IRQ_HANDLER(n)
```

Comment:
GD32_USART_IRQ_HANDLER(n)` ; `

Reply:
Since GD32_USART_IRQ_HANDLER(n) actually is a function definition, ';' are not require in here, that's why push it to the end.